beam-sync.md

```diff
@@ -0,0 +1,187 @@
+
+### State Retrieval Performance
+
+Naively, when handling a data fault, each `GetNodeData` request can only contain a single trie node. Starting from the empty trie, you can download the root node, but you can't download any other nodes on the path to the account you care about yet, because they are all hash-indexed. The hash of the child is in the body of the parent. So, proving the balance of a single account with an empty trie takes as many requests as there are levels to the trie (say ~7 in mainnet, currently). 
+
+So performance is highly sensitive to latency: round-trip time to your best peer is the dominating factor of how quickly you can execute the EVM with JIT data.
+
+### Lag
+
+It can take longer than a block time to collect all the state needed for a block. So as new blocks arrive,
+the currently executing block can fall further and further behind the tip. Clients typically
+only serve state for recent blocks, configured differently for different clients. A
+common number is ~100 blocks of state retained. If you fall behind this, then missing state
+may no longer be available to you. This can leave you indefinitely stuck, so when we
+lag far enough behind, we pivot.
+
+### Pivoting
+
+Pivoting skips executing the current blocks, and chooses the tip block again. It will
+typically increase the missing state that must be backfilled. It is quite likely that Beam Sync
+v0 will require a few pivots before it exits. As it progresses, and stores more state, pivoting gets less likely.
+v1 and v2 should be fast enough that pivoting is just a failsafe against "attack blocks" with
+intentionally gigantic witnesses.
+
+### Optimizations
+
+A naive implementation of Beam Sync, given that there are on the order of 4k trie nodes read per mainnet block, and a good round-trip time is ~100ms, it would take 400s to collect all the required data to execute a single block. With ~15-second block times, this is clearly a problem.
+
+So while optimizations are not *technically* required, they are necessary in practice to avoid getting stuck in a loop of pivots. A pivot loop would mean never catching up to complete Full Sync.
+
+Below are some optimizations that Trinity has found effective. You may notice the theme that they all provide ways to look up more data in parallel. We would much rather be limited by total peer bandwidth across all connected peers, rather than limited by the round-trip latency of a single peer.
+
+#### Optimization 1: Address Batching
+
+Before starting EVM execution, we can identify several key addresses that will be needed during the block: transaction recipients, transaction senders, and the header's coinbase.
```

Comment:
Did you intentionally exclude the logs from receipts or did we find these weren't useful?

Reply:
Sure, I can mention it as "unexplored." The network cost and implementation time is a little higher, since it's the only reason we would download block receipts at all. I have a weak prediction that the number of extra addresses we could extract this way is small compared to the number of other addresses we collect as mentioned in the doc. So it's on the backlog to explore whether it's worth it, but not at the top, at the moment. (Witness metadata propagation being much higher)